File: go/excercise318/maximum_product_of_word_lengths.go
package maximum_product_of_word_lengths

func maximumProductOfWordLengths(words []string) int {
	n := len(words)
	bits := make([]int, n)
	ans := 0
	for i := 0; i < n; i++ {
		for _, char := range words[i] {
			bits[i] |= 1 << (char - 'a')
		}

		for j := 0; j < i; j++ {
			if bits[i]&bits[j] == 0 {
				prod := len(words[i]) * len(words[j])
				if prod > ans {
					ans = prod
				}
			}
		}
	}
	return ans
}

File: python/excercise318/maximum_product_of_word_lengths.py
from typing import List


def maximum_product_of_word_lengths(words: List[str]) -> int:
    n = len(words)
    bits = [0] * n

    ans = 0
    for i in range(n):
        for ch in words[i]:
            bits[i] |= 1 << (ord(ch) - ord('a'))
        for j in range(0, i):
            if bits[i] & bits[j] == 0:
                prod = len(words[i]) * len(words[j])
                ans = max(ans, prod)
    return ans
